csrc/ops/utils.cpp

```diff
@@ -314,7 +314,7 @@ Val* newValLike(Val* val, DataType dtype) {
     return newOutputTV({val}, dtype);
   }
 
-  return newScalar(vtype, dtype);
+  return newScalar(ValType::Others, dtype);
```

Comment:
Why this change?

Reply:
Otherwise we will be unable to do things like `maybeCastOp(TensorIndex)`

Reply:
Sorry, I'm lost here. Why is it? Is it documented somewhere?

Reply:
I don't think it is documented anywhere. It is just because the results of ops on `TensorIndex` are no longer a `TensorIndex` but a general `Val`. So we should not always preserve the same `ValType` here.

Reply:
OK, now it makes sense. I'd say ops on `TensorIndex` have been `ValType::Others`. It's just we were not using `maybeCastOp` before but manually created a new Val of type `ValType::Others`. 

Part of the reason I feel this is a little odd would be probably due to `TensorIndex`. I remember you mentioned that it should be an expression rather than a val, which I agree.